### DISCUSSION TITLE
Make association properties can be subclass of types implemented IEnumberable<T>

### DISCUSSION THREAD

**Discussion by Ethen-Wang** [linq2db/linq2db#4351] [General]
We are planning to use linq2db as the ORM framework to replace the current one which developed by ourselves. In the current code all the association properties are inherit from a super class which implements IEnumberable<T>. This is an example of our code and all the attributes are defined by ourselves. 
```
[Table(Name = "Carrier")]
public class Carrier
{
    [Column(Name = "InternalID", DataType = SqlDbType.Int, PrimaryKey = true, AllowNull = false)]
    public int InternalID { get; set; }

    [Relationship(ForeignKey = "SCAC", ParentKey = "SCAC", Kind = RelationshipType.OneToMany)]
    public BillingSettingsList CarrierEngineAccountLinkList { get; set; }
}

[Table(Name = "CarrierAccountBilling")]
public class BillingSettingsList : TransactionObjectArray<CarrierAccountBilling> // implements IEnumberable<CarrierAccountBilling>
{
}
```
we are trying to use Fluent Mapping to build schema maps. All the things work fine except the "BillingSettingsList" property. linq2db just cannot find the element type for this property. I checked in the source code and seems this can be easily fixed by updating GetEnumerableElementType method in EagerLoading.cs to be like following. 
```
public static Type GetEnumerableElementType(Type type, MappingSchema mappingSchema)
{
	if (!IsEnumerableType(type, mappingSchema))
		return type;
	if (type.IsArray)
		return type.GetElementType()!;
	if (typeof(IGrouping<,>).IsSameOrParentOf(type))
		return type.GetGenericArguments()[1];
	return type.GetGenericArguments(typeof(IEnumerable<>))![0]; // add type parameter
}
```
Can it be included in future release or any other solutions for this problem?

**Comment by sdanyliv** [comment]
Problem also is how to tell `linq2db` how instantiate `BillingSettingsList`.

**Reply by Ethen-Wang** [reply]
We are using the mapping converter to convert from `IEnumerable<CarrierAccountBilling>` to BillingSettingsList

**Comment by sdanyliv** [comment]
Anyway, Pull request with tests are welcome, we will correct if something can be improved.

**Reply by Ethen-Wang** [reply]
With the updated code both` IEnumerable<CarrierAccountBilling>` and BillingSettingsList work fine. I don't have the permission to make PR to this repo.

**Reply by sdanyliv** [reply]
1. Make fork of `linq2db` repository. 
2. Create in your `linq2db` github repository branch. 
3. Make changes.
4. Create pull request.

[Creating a pull request from a fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork)

**Comment by Ethen-Wang** [comment]
Hi,

Thanks for replying. And can I have one more question? In our code we have a calculate field as the foreign key. How can we tell linq2db to take this field into consideration in queries? From my researching ExpressionMethodAttribute might be the answer but not sure how to make it work. Can you give us any suggestions or examples to handle this scenario?
```
[Table(Name = "Carrier")]
public class Carrier
{
    [Column(Name = "InternalID", DataType = SqlDbType.Int, PrimaryKey = true, AllowNull = false)]
    public int InternalID { get; set; }

    [XmlAttribute("LocationID")]
    [Column(Name = "LocationID", DataType = SqlDbType.Int, AllowNull = true)]
    public int LocationID { get; set; }

    [XmlAttribute("SCAC")]
    [Column(Name = "SCAC", DataType = SqlDbType.VarChar, Length = 5)]
    public string SCAC { get; set; }

    [XmlAttribute("CarrierLink")]
    public string CarrierLink => $"{SCAC}{(LocationID > 0 ? "_" + LocationID.ToString() : "")}"; // calculate field as a foreign key

    [XmlElement("BillingSettings")]
    [Relationship(ForeignKey = "CarrierLink", ParentKey = "CarrierLink", Kind = RelationshipType.OneToMany)]
    public BillingSettingsList CarrierEngineAccountLinkList { get; set; }
}

[Table(Name = "CarrierAccountBilling")]
public class BillingSettingsList : TransactionObjectArray<CarrierAccountBilling> // implements IEnumberable<CarrierAccountBilling>
{
    public string CarrierLink { get; set; }
}
```
